pkg/operator/controller/configurable-route/controller.go

```diff
@@ -0,0 +1,402 @@
+	for _, componentRoute := range componentRouteStatuses {
+		componentRouteHash := util.Hash(namespacedName(componentRoute.Namespace, componentRoute.Name))
+		componentRouteHashToComponentRouteStatus[componentRouteHash] = componentRoute
+	}
+
+	componentRoutes := []aggregatedComponentRoute{}
+	for _, componentRouteSpec := range componentRouteSpecs {
+		hash := util.Hash(namespacedName(componentRouteSpec.Namespace, componentRouteSpec.Name))
+		if componentRouteStatus, ok := componentRouteHashToComponentRouteStatus[hash]; ok {
+			componentRoutes = append(componentRoutes, newAggregatedComponentRoute(componentRouteSpec, componentRouteStatus))
+		}
+	}
+	return componentRoutes
+}
+
+func namespacedName(namespace, name string) string {
+	return fmt.Sprintf("%s/%s", namespace, name)
+}
+
+func componentRouteResources(componentRoute aggregatedComponentRoute) []client.ListOption {
+	return []client.ListOption{
+		client.MatchingLabels{
+			componentRouteHashAnnotation: componentRoute.Hash,
+		},
+		client.InNamespace(operatorcontroller.OpenShiftConfigNamespace),
+	}
+}
+
+func allComponentRouteResources() []client.ListOption {
+	return []client.ListOption{
+		client.HasLabels{componentRouteHashAnnotation},
+		client.InNamespace(operatorcontroller.OpenShiftConfigNamespace),
+	}
+}
+
+func (r *reconciler) deleteOrphanedRoles(componentRoutes []aggregatedComponentRoute) error {
+	existingHashes := map[string]struct{}{}
+	for _, cr := range componentRoutes {
+		existingHashes[cr.Hash] = struct{}{}
+	}
+
+	roleBindingList := &rbacv1.RoleBindingList{}
+	r.cache.List(context.TODO(), roleBindingList, allComponentRouteResources()...)
+	for _, item := range roleBindingList.Items {
+		expectedHash, ok := item.GetLabels()[componentRouteHashAnnotation]
+		if !ok {
+			return fmt.Errorf("Unable to find expected componentRoute hash label")
+		}
+
+		if _, ok := existingHashes[expectedHash]; !ok {
+			log.Info("deleting roleBinding", "name", item.GetName(), "namespace", item.GetNamespace())
+			if err := r.client.Delete(context.TODO(), &item); err != nil && !errors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *reconciler) deleteOrphanedRoleBindings(componentRoutes []aggregatedComponentRoute) error {
+	existingHashes := map[string]struct{}{}
+	for _, cr := range componentRoutes {
+		existingHashes[cr.Hash] = struct{}{}
+	}
+
+	roleBindingList := &rbacv1.RoleBindingList{}
+	r.cache.List(context.TODO(), roleBindingList, allComponentRouteResources()...)
+	for _, item := range roleBindingList.Items {
+		expectedHash, ok := item.GetLabels()[componentRouteHashAnnotation]
+		if !ok {
+			return fmt.Errorf("Unable to find expected componentRoute hash label")
+		}
+
+		if _, ok := existingHashes[expectedHash]; !ok {
+			log.Info("deleting roleBinding", "name", item.GetName(), "namespace", item.GetNamespace())
+			if err := r.client.Delete(context.TODO(), &item); err != nil && !errors.IsNotFound(err) {
```

Comment:
are we sure here this does not mutate the cache, i.e. when item has some pointer substruct. Won't they be reused by the delete client?